### PR TITLE
HackerNewsCorpus and LdaMallet model

### PR DESCRIPTION
**Pull request by WGierke**
I added a class for tokenizing comments of the HackerNews dataset and a LDA model that is based on Java MALLET.
The found 10 topics in the comments of `HN_2017-07` represented by their top 20 words are:
```python
# 0	5	don data service doesn problem security system access bitcoin case network internet information phone account gt services key number
# 1	5	google web open app source windows don facebook content pretty support game apple users apps site page free works
# 2	5	code language pre data type write run system python languages js performance programming function memory simple java tools api
# 3	5	https href rel nofollow www http org news github en wiki wikipedia amp html id ycombinator item blog youtube
# 4	5	people don person world social law government women fact china bad society country state evidence opinion legal political situation
# 5	5	quot gt don point doesn read understand article isn question true comment wrong make sense thing idea answer makes
# 6	5	problem power space human change long point world system ai high large energy scale current speed problems isn low
# 7	5	work company good software time working experience job product full tech make great team people jobs companies business end
# 8	5	time don ve years things people good day back thing didn lot ll long bad feel days life ago
# 9	5	money pay market cost people year buy car price high big live companies costs business worth cars area make

 # <1000> LL/token: -8.76975 (took 12 minutes)
```
Note how topic 2 is about programming, 3 represents parts of links, 4 is about social and political stuff, 7 is about problems and 9 about businesses and money.

### CODE REVIEW

File: src/models/corpus.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def iter_documents(documents):
+    """Iterate over comments, yielding one document at a time."""
+    for document in documents:
+        # parse document into a list of utf8 tokens
+        yield utils.simple_preprocess(document)
+
+
+class HackernewsCorpus(object):
```
Comment (mockingbird2) [inline src/models/corpus.py]: Can we agree on using python3? Since you can use it without the object inheritance

File: src/models/training.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    # Focus on the stories
+    comments_df = df[df.type == "comment"]
+    comments_df = comments_df.dropna(subset=['text'])
+    comments = comments_df['text']
+
+    # Set up the streamed corpus
+    corpus = HackernewsCorpus(comments)
+
+    # Train 10 LDA topics using MALLET
+    mallet_path = '/home/madness/Programs/mallet-2.0.6/bin/mallet'
```
Comment (mockingbird2) [inline src/models/training.py]: Can you add installation instructions for mallet, please?

File: src/models/training.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    # 8	5	time don ve years things people good day back thing didn lot ll long bad feel days life ago
+    # 9	5	money pay market cost people year buy car price high big live companies costs business worth cars area make
+
+    # <1000> LL/token: -8.76975
+
+    # Note how topic 2 is about programming, 3 represents parts of links, 4 is about social and political stuff,
+    # 7 is about problems and 9 about businesses and money
+
+    # Now use the trained model to infer topics on a new document
+    doc = "Deep Learning is great"
```
Comment (mockingbird2) [inline src/models/training.py]: Is this sentence really representative? It seems to me this file should be rather a notebook.

File: src/models/training.py (hunk 2 above)
Comment (fawind) [inline src/models/training.py]: Can you make this hardcoded path relative? I'm guessing not everyone is named Madness ;)

Comment (WGierke) [reply src/models/training.py]: So rename yourself :)

**Comment by WGierke** [comment]
@mockingbird2 How about now?
My initial plan was to use MALLET as LDA backend and visualize the topics using pyLDAvis so we can easier interpret the topics. The expected output was [something like this](http://nbviewer.jupyter.org/github/bmabey/hacker_news_topic_modelling/blob/master/HN%20Topic%20Model%20Talk.ipynb#topic=0&lambda=1&term=). However, it turns out that when converting the MALLET model back to a gensim LDA model, the learned topics aren't copied meaning the model needs to be trained again ... I'd still like to merge this branch into the master as it adds a working topic modeling approach and a nice corpus wrapper.

**Comment by mockingbird2** [comment]
Are you using python2? When I try to run the topic modelling notebook I get encoding errors?

**Comment by mockingbird2** [comment]
@WGierke Have you tried this [gensim lda](https://radimrehurek.com/gensim/models/ldamallet.html)?

**Pull request merged**